coderd/httpapi/name.go

```diff
@@ -35,16 +22,47 @@ func UsernameValid(str string) error {
 // the username from an email address. If no success happens during
 // these steps, a random username will be returned.
 func UsernameFrom(str string) string {
-	if valid := UsernameValid(str); valid == nil {
+	if valid := NameValid(str); valid == nil {
 		return str
 	}
 	emailAt := strings.LastIndex(str, "@")
 	if emailAt >= 0 {
 		str = str[:emailAt]
 	}
 	str = usernameReplace.ReplaceAllString(str, "")
-	if valid := UsernameValid(str); valid == nil {
+	if valid := NameValid(str); valid == nil {
 		return str
 	}
 	return strings.ReplaceAll(namesgenerator.GetRandomName(1), "_", "-")
 }
+
+// NameValid returns whether the input string is a valid name.
+// It is a generic validator for any name (user, workspace, template, etc.).
+func NameValid(str string) error {
+	if len(str) > 32 {
+		return xerrors.New("must be <= 32 characters")
+	}
+	if len(str) < 1 {
+		return xerrors.New("must be >= 1 character")
+	}
+	matched := UsernameValidRegex.MatchString(str)
+	if !matched {
+		return xerrors.New("must be alphanumeric with hyphens")
+	}
+	return nil
+}
+
+// TemplateDisplayNameValid returns whether the input string is a valid template display name.
+func TemplateDisplayNameValid(str string) error {
+	if len(str) == 0 {
+		return nil // empty display_name is correct
+	}
+	if len(str) > 32 {
+		return xerrors.New("must be <= 32 characters")
```

Comment:
Database had a `VARCHAR(64)` is this the intended value?

Also, do we actually need to limit the length?

Reply:
> Database had a VARCHAR(64) is this the intended value?

Good question. I admit that I based on the `template.name` here, which is defined as `name character varying(64) NOT NULL` in `dump.sql`, but the validator limits it to 32 chars.

I can fix it in both places, but not sure about the implications. I guess that 64 chars were the older limit?

> Also, do we actually need to limit the length?

As I replied to the other comment. As the property is intended to replace `template.name` in the UI, I would set it to a reasonable value.

Reply:
The `name`s have limited length (32) due to use in URLs etc, I don't think we should inherit that for a human readable display name. Just my full name alone (middle names included) is 31 characters, barely making the cut if I wanted to name a template for myself 😄. The UI can always use a text ellipse or similar when names are too long.

Also, historically (in this repo) there has been very little coordination between database limits and API limits (which is bad from both a UX and understanding the code perspective), we basically have two options: 1) Keep limits/checks in sync between API and database or 2) remove them from one or the other.

Personally I'm a fan of the database being a "last line of defense", and as such like having the check in both places. But this is harder to keep right maintainability wise and some checks are hard(er) to do in the database, so we may end up with discrepancy either way.

Reply:
> Personally I'm a fan of the database being a "last line of defense", and as such like having the check in both places. 

Yes, that's the approach I'm thinking about too. Let's set it to 64 characters for now. We can always increase it if it's insufficient.